msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java

```diff
@@ -42,12 +42,17 @@
     private final String mTenantId;
     private final AccessTokenRecord mAccessToken;
     private final IAccount mAccount;
+    private boolean mServicedFromCache;
 
-    AuthenticationResult(@NonNull final List<ICacheRecord> cacheRecords) {
+    AuthenticationResult(
+            @NonNull final List<ICacheRecord> cacheRecords,
+            final boolean isServicedFromCache
```

Comment:
This is just about whether this is served from the local MSAL cache, I think.  Are we going to be similarly curious if this was served by an auth broker?

Reply:
Yes, this is the original work item: https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/875729

I split it into two legs: served from cache, served from broker.

This the work item I created for implementing the served from broker: https://dev.azure.com/IdentityDivision/Engineering/_workitems/edit/1022834

Going to be working on that next!

Reply:
OK.  There's not a lot of detail in these work items in terms of the actual desire here.  This makes me wonder: should this be a boolean or should it be some kind of object that could contain more information.  Because the next questions would be "which broker", and "how long has it been in this cache", "was it cached by the broker involved", etc.  I think we (should) get the age of the decision from the auth token itself, I think.

Reply:
I think we should definitely do "which broker". I think if the broker was being used, then it is implied that when the result was serviced from cache then it was serviced from the cache of the broker. For specifying how long it's been there would be a bit of an overkill IMO but open to suggestions on that.

Reply:
Right, I keep forgetting that the MSAL cache isn't a first tier before the broker process, though some form of sharing might be desirable in the longer term.  Generally, though, knowing the age of objects in a cache is useful for monitoring its effectiveness (and finding bugs in age-related code), though I think we don't do capacity-based cache eviction.

The discussion of record age aside, if you just had a source object with a few different potential kinds of value, maybe

local
broker(identity-of-broker)
authority(identity-of-authority?)

that would let you put all of this information into a single method and save us trouble in the future if we needed to add more of them, unless there's some reason to strongly prefer a primitive type in the return.

Reply:
@AdamBJohnsonx We could create object called as `ResultMetadata` and put these values like whether the result was served from cache or broker, which broker was used etc. inside that object and then expose that to the developers. 

For now we would have 3 fields: `wasServicedFromCache`, `wasServicedFromBroker`, `brokerUsed`, but as you said, we may want to add more fields here going forward so I think having a metadata object makes sense. Since this public API, I am holding off on merging this, let's think more about and let me know if you have any suggestions.

Reply:
This is actually a separate question - this API is about more than just Android, so if we're already committed to a common API where these are booleans and strings then my entire discussion is moot. ;-)

Reply:
This API is specifically for Android. AFAIK other platforms have not implemented this. This was something that was asked by the CSS support team to enable in Android. @iambmelt would have more context on that.

I think if other platforms choose to implement such a meta data, then they need to use the return types applicable to that platform/language.

Reply:
Then in this case, having a containing type makes the most sense for us.  It does for other MSALs as well, honestly, since it's a stumbling block for anybody working on or using an auth client, and hopefully the response shape is something that we can align on with other clients.  Developers on android are occasionally the same people working on iOS, etc. and it's frustrating when I can't use the library for one system in the same way I can the second.

Reply:
@AdamBJohnsonx If we are going to doing that then we should probably do that across all public apis in Android and iOS. That is not the case right now. For example, the errors returned by iOS and Android via their public api are completely different..they are even completely different types. Android exposes errors as exception objects and error codes as Strings whereas iOS does so by using enums: https://github.com/AzureAD/microsoft-authentication-library-for-objc/blob/master/MSAL/src/public/MSALError.h#L128

TBH, I don't even know what types we may be able to use that are the same across platforms. IMHO, when creating public API and selecting types, we should be picking types that offer the best experience on a given platform.

Reply:
If we haven't done it it isn't worth limiting ourselves with it.  I don't think we're disagreeing, mind, I was more making a note that this feature is surely going to be desired elsewhere.